### ISSUE TITLE
dqlite issues preventing startup after reboot

### ISSUE THREAD

**Issue by ericrochow** [lxc/lxd#6917]
# Required information

 * Distribution: Ubuntu
 * Distribution version: 18.04
 * The output of "lxc info" or if that fails:
   * Kernel version: 4.15.0-88-generic
   * LXC version: 3.20 (snap)
   * LXD version: 3.20 (snap)
   * Storage backend in use: ZFS

# Issue description

Following an unclean reboot of my host (console was hung and had to forcably reboot), I noticed that the containers running on it were no longer assessible following a reboot. I attempted to view the status of my containers, but every `lxc` command returns the following:

```
Error: Get http://unix.socket/1.0: dial unix /var/snap/lxd/common/lxd/unix.socket: connect: connection refused
```

# Steps to reproduce

Not sure how to reproduce, just cropped up after a reboot.

# Information to attach

dmesg throws a lot of the following:

```
[ 1392.539663] new mount options do not match the existing superblock, will be ignored
[ 1394.238071] new mount options do not match the existing superblock, will be ignored
[ 1395.764771] new mount options do not match the existing superblock, will be ignored
[ 1397.266084] new mount options do not match the existing superblock, will be ignored
[ 1492.104663] new mount options do not match the existing superblock, will be ignored
[ 1493.739143] new mount options do not match the existing superblock, will be ignored
[ 1495.250561] new mount options do not match the existing superblock, will be ignored
```

The main daemon log contains the following:

```
❯ sudo tail -50 /var/snap/lxd/common/lxd/logs/lxd.log
t=2020-02-21T19:26:48-0500 lvl=info msg="LXD 3.20 is starting in normal mode" path=/var/snap/lxd/common/lxd
t=2020-02-21T19:26:48-0500 lvl=info msg="Kernel uid/gid map:" 
t=2020-02-21T19:26:48-0500 lvl=info msg=" - u 0 0 4294967295" 
t=2020-02-21T19:26:48-0500 lvl=info msg=" - g 0 0 4294967295" 
t=2020-02-21T19:26:48-0500 lvl=info msg="Configured LXD uid/gid map:" 
t=2020-02-21T19:26:48-0500 lvl=info msg=" - u 0 1000000 1000000000" 
t=2020-02-21T19:26:48-0500 lvl=info msg=" - g 0 1000000 1000000000" 
t=2020-02-21T19:26:48-0500 lvl=info msg="Kernel features:" 
t=2020-02-21T19:26:48-0500 lvl=info msg=" - netnsid-based network retrieval: no" 
t=2020-02-21T19:26:48-0500 lvl=info msg=" - uevent injection: no" 
t=2020-02-21T19:26:48-0500 lvl=info msg=" - seccomp listener: no" 
t=2020-02-21T19:26:48-0500 lvl=info msg=" - seccomp listener continue syscalls: no" 
t=2020-02-21T19:26:48-0500 lvl=info msg=" - unprivileged file capabilities: yes" 
t=2020-02-21T19:26:48-0500 lvl=info msg=" - cgroup layout: hybrid" 
t=2020-02-21T19:26:48-0500 lvl=warn msg=" - Couldn't find the CGroup memory swap accounting, swap limits will be ignored" 
t=2020-02-21T19:26:48-0500 lvl=info msg=" - shiftfs support: disabled" 
t=2020-02-21T19:26:48-0500 lvl=info msg="Initializing local database" 
t=2020-02-21T19:26:49-0500 lvl=eror msg="Failed to start the daemon: Failed to start dqlite server: raft_start(): io: load closed segment 0000000000039733-0000000000040071: found 338 entries (expected 339)" 
t=2020-02-21T19:26:49-0500 lvl=info msg="Starting shutdown sequence" 
```

Manual start with debugging:

```❯ sudo lxd --debug --group lxd
DBUG[02-21|19:30:32] Connecting to a local LXD over a Unix socket 
DBUG[02-21|19:30:32] Sending request to LXD                   method=GET url=http://unix.socket/1.0 etag=
INFO[02-21|19:30:32] LXD 3.20 is starting in normal mode      path=/var/snap/lxd/common/lxd
INFO[02-21|19:30:32] Kernel uid/gid map: 
INFO[02-21|19:30:32]  - u 0 0 4294967295 
INFO[02-21|19:30:32]  - g 0 0 4294967295 
INFO[02-21|19:30:32] Configured LXD uid/gid map: 
INFO[02-21|19:30:32]  - u 0 1000000 1000000000 
INFO[02-21|19:30:32]  - g 0 1000000 1000000000 
INFO[02-21|19:30:32] Kernel features: 
INFO[02-21|19:30:32]  - netnsid-based network retrieval: no 
INFO[02-21|19:30:32]  - uevent injection: no 
INFO[02-21|19:30:32]  - seccomp listener: no 
INFO[02-21|19:30:32]  - seccomp listener continue syscalls: no 
INFO[02-21|19:30:32]  - unprivileged file capabilities: yes 
INFO[02-21|19:30:32]  - cgroup layout: hybrid 
WARN[02-21|19:30:32]  - Couldn't find the CGroup memory swap accounting, swap limits will be ignored 
INFO[02-21|19:30:32]  - shiftfs support: no 
INFO[02-21|19:30:32] Initializing local database 
DBUG[02-21|19:30:32] Initializing database gateway 
DBUG[02-21|19:30:32] Start database node                      id=1 address= role=voter
EROR[02-21|19:30:33] Failed to start the daemon: Failed to start dqlite server: raft_start(): io: load closed segment 0000000000039733-0000000000040071: found 338 entries (expected 339) 
INFO[02-21|19:30:33] Starting shutdown sequence 
DBUG[02-21|19:30:33] Not unmounting temporary filesystems (containers are still running) 
Error: Failed to start dqlite server: raft_start(): io: load closed segment 0000000000039733-0000000000040071: found 338 entries (expected 339)
```

Let me know if there is anything else I can provide.

**Comment by stgraber** [comment]
@ericrochow can you run `snap refresh lxd` and try again?
We landed a fix for this issue this morning.

**Comment by freeekanayaka** [comment]
I think this is actually a different issue, probably some corruption happened because of the crash.

Please make a backup copy of your ```/var/snap/lxd/common/lxd/database``` directory, then run

```
rm /var/snap/lxd/common/lxd/database/global/0000000000039733-0000000000040071
```

and try to start the daemon again.

If it starts ok but you notice anything missing (containers or whatever), just let me know and we can recover them.

**Comment by ericrochow** [comment]
I tried the snap refresh and that did not recover. I moved the requested file to my home directory and was able to start and it looks like I'm back in business. Is there a way of determining what was in that file? The last change I made to anything was to detach a container from a profile and attach to a new one, but that would have been a few days ago. From what I can see now, it appears that change persisted. Would this have been some background operation?
